bee-pow/src/providers/miner.rs

```diff
@@ -0,0 +1,146 @@
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread,
+};
+
+const DEFAULT_NUM_WORKERS: usize = 1;
+/// https://oeis.org/A002391
+const LN_3: f64 = 1.098_612_288_668_109_8;
+
+/// Errors occurring when computing nonces with the `Miner` nonce provider.
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("The worker has been cancelled.")]
+    Cancelled,
+}
+
+/// Builder for the `Miner` nonce provider.
+#[derive(Default)]
+pub struct MinerBuilder {
+    num_workers: Option<usize>,
+}
+
+impl MinerBuilder {
+    /// Sets the desired number of workers for the `Miner` nonce provider.
+    pub fn with_num_workers(mut self, num_workers: usize) -> Self {
+        self.num_workers = Some(num_workers);
+        self
+    }
+}
+
+impl NonceProviderBuilder for MinerBuilder {
+    type Provider = Miner;
+
+    fn new() -> Self {
+        Self::default()
+    }
+
+    fn finish(self) -> Miner {
+        Miner {
+            num_workers: self.num_workers.unwrap_or(DEFAULT_NUM_WORKERS),
+        }
+    }
+}
+
+/// A nonce provider that mine nonces.
+pub struct Miner {
+    num_workers: usize,
+}
+
+impl Miner {
+    fn worker(
+        done: Arc<AtomicBool>,
+        pow_digest: TritBuf<T1B1Buf>,
+        start_nonce: u64,
+        target_zeros: usize,
+    ) -> Result<u64, Error> {
+        assert!(target_zeros <= HASH_LENGTH);
+
+        let mut nonce = start_nonce;
+        let mut hasher = BatchHasher::<T1B1Buf>::new(HASH_LENGTH, CurlPRounds::Rounds81);
+        let mut buffers = Vec::<TritBuf<T1B1Buf>>::with_capacity(BATCH_SIZE);
+
+        for _ in 0..BATCH_SIZE {
+            let mut buffer = TritBuf::<T1B1Buf>::zeros(HASH_LENGTH);
+            buffer[..pow_digest.len()].copy_from(&pow_digest);
+            buffers.push(buffer);
+        }
+
+        while !done.load(Ordering::Relaxed) {
+            for (i, buffer) in buffers.iter_mut().enumerate() {
+                let nonce_trits = b1t6::encode::<T1B1Buf>(&(nonce + i as u64).to_le_bytes());
+                buffer[pow_digest.len()..pow_digest.len() + nonce_trits.len()].copy_from(&nonce_trits);
+                hasher.add(buffer.clone());
+            }
+
+            for (i, hash) in hasher.hash_batched().enumerate() {
+                let trailing_zeros = hash.iter().rev().take_while(|t| *t == Btrit::Zero).count();
+
+                if trailing_zeros >= target_zeros {
+                    done.store(true, Ordering::Relaxed);
+                    return Ok(nonce + i as u64);
+                }
+            }
+
+            nonce += BATCH_SIZE as u64;
+        }
+
+        Err(Error::Cancelled)
+    }
+}
+
+impl NonceProvider for Miner {
+    type Builder = MinerBuilder;
+    type Error = Error;
+
+    fn nonce(&self, bytes: &[u8], target_score: f64, done: Option<Arc<AtomicBool>>) -> Result<u64, Self::Error> {
+        let mut nonce = 0;
+        let mut pow_digest = TritBuf::<T1B1Buf>::new();
+        let done = done.unwrap_or_else(|| Arc::new(AtomicBool::new(false)));
```

Comment:
See my proposal regarding the `Arc<AtomicBool>`. I think it would be a better design if the `Arc<AtomicBool>` would be created by the builder. We then wouldn't need to create it here, if the user didn't hand us one.